kcidb/__init__.py

```diff
@@ -407,3 +412,95 @@ def ingest_main():
             )
             sys.stdout.write("\x00")
             sys.stdout.flush()
+
+
+def triage_main():
+    """Execute the kcidb-triage command-line tool"""
+    sys.excepthook = misc.log_and_print_excepthook
+    description = 'kcidb-triage - Triage builds/tests against issues in ' \
+        'the specified object set'
+    parser = oo.ArgumentParser(database="json", description=description)
+    parser.add_argument(
+        '-b', '--cache-bucket',
+        help='Name of the cache bucket to try fetching artifacts from. '
+        'The value of KCIDB_CACHE_BUCKET_NAME is used, if not specified. '
+        'Otherwise artifacts are downloaded from the original URLs.'
+    )
+    args = parser.parse_args()
+    # Get the objects to be triaged
+    pattern_set = set()
+    for pattern_string in args.pattern_strings:
+        pattern_set |= orm.query.Pattern.parse(pattern_string)
+    oo_client = oo.Client(db.Client(args.database))
+    objects = oo_client.query(pattern_set)
+    # Create the cache client
+    cache_bucket = args.cache_bucket or \
+        os.environ.get("KCIDB_CACHE_BUCKET_NAME")
+    cache_client = cache_bucket and cache.Client(cache_bucket, 0, 0)
+    cache_ttl = datetime.timedelta(seconds=30)
+    requests_session = requests.Session()
+    output = io.SCHEMA.new()
+    output_incidents = output.setdefault("incidents", [])
+
+    def fetch_log(obj):
+        """Fetch the text of the main log for an object"""
+        if not obj.log_url:
+            return None
+        log_url = (
+            cache_client and
+            cache_client.map(obj.log_url, cache_ttl)
+        ) or obj.log_url
+        try:
+            LOGGER.debug("Downloading log for %s ID %r from %r",
+                         type(obj).__name__, obj.get_id(), log_url)
+            response = requests_session.get(log_url,
+                                            timeout=30,
+                                            allow_redirects=True)
```

Comment:
I strongly recommend limiting the download size from the start.
It can avoid unexpected memory usage problems, e.g. CKI has recently submitted a build with a log of 320MB that has caused some problems.

```suggestion
            response = requests_session.get(log_url,
                                            timeout=30,
                                            stream=True,
                                            allow_redirects=True)
            content_iterator = response.iter_content(DOWNLOAD_MAX_SIZE)
            if not (raw_data := next(content_iterator, None)):
                return None
```

Reply:
Of course. We have that implemented in the artifact archive. I could just copy it from there. This is just a PoC.